bika/lims/browser/workflow/client.py

```diff
@@ -0,0 +1,74 @@
+# -*- coding: utf-8 -*-
+
+import itertools
+
+from bika.lims import _
+from bika.lims import api
+from bika.lims import logger
+from bika.lims.browser.workflow import RequestContextAware
+from bika.lims.interfaces import IWorkflowActionUIDsAdapter
+from Products.CMFCore.WorkflowCore import WorkflowException
+from zope.component.interfaces import implements
+
+
+class WorkflowActionPublishSamplesAdapter(RequestContextAware):
+    """Adapter in charge of the client 'publish_samples' action
+    """
+    implements(IWorkflowActionUIDsAdapter)
+
+    def __call__(self, action, uids):
+        published = []
+
+        # get the selected ARReport objects
+        reports = map(api.get_object_by_uid, uids)
+        # get all the contained sample UIDs of the generated PDFs
+        sample_uids = map(self.get_sample_uids_in_report, reports)
+        # uniquify the UIDs of the contained samples
+        unique_sample_uids = set(list(
+            itertools.chain.from_iterable(sample_uids)))
```

Comment:
Is there any reason why you use `itertools` here?. I guess you could do `set(sample_uids)`  instead

Reply:
The `sample_uids` is actually a list of lists.
For better understanding:
- The users selects 1..n ARReports from the list
- Each report might be a multi-report containing more than one AR/UID
- Of all the selected ARReports, the UIDs need to be extracted -> list of UID lists
- All of these UIDs must be uniquified, looked up and finally transitioned

Reply:
Aa, I see, thanks for the clarification